### DISCUSSION TITLE
World mode

### DISCUSSION THREAD

**Discussion by brenocq** [brenocq/implot3d#17] [General]
### This discussion was converted into issue #106

ImPlot3D currently confines all plots to a **box-shaped bounded space**, which is great for most data visualizations. However, there are cases where a more **open and exploratory environment** would make much more sense — where the space isn't limited by a box. To address this, I'd like to propose two distinct modes:

## Box Mode (Current Implementation)
- The plot is enclosed within a bounding box, with fixed axis limits.
- Navigation is limited to rotating, zooming, and panning within the box.
- This is perfect for datasets that fit well in defined bounds or when grid-aligned visualizations are what you need.

## World Mode (New Proposal)
- The bounding box is removed, and instead, we'd have an **infinite ground plane** on the XY plane as a visual reference.
- Users could freely move, pan, and rotate the camera, similar to navigating 3D environments in tools like Blender or Unity.
- **Where it would shine**:
  - Visualizing massive or unbounded datasets (e.g., GIS data or simulations).
  - Exploring spatial data interactively.
  - Building immersive, world-like visualizations.

---

### Proposed Features of World Mode
1. **Infinite Ground Plane**:
   - Expand XY plane grid to be unbounded and hide the YZ/XZ grids.

2. **Free Camera Navigation**:
   - First-person or free-camera controls for unrestricted exploration.
   - Keyboard/mouse controls (e.g., WASD for movement, mouse for looking around).
   - Support for **orthographic and perspective projections**:

3. **(Future Optimization)**: Avoid rendering meshes outside the camera's field of view. (But this is a feature on its own 😅).

---

### How It Could Work
I believe we can implement this by only changing the BeginPlot to

```cpp
if (ImPlot3D::BeginPlot("World Mode Example", ImVec2(-1, 200), ImPlot3DFlags_WorldMode) {
    ImPlot3D::PlotScatter("Data Points", x_data, y_data, z_data, count);
    ImPlot3D::PlotMesh("3D Object", vertices, indices, vertex_count, index_count);
    ImPlot3D::EndPlot();
}
```

---

This feature is a bit of a wild card, but I think it would be super useful for certain applications (I'd love to use it myself 😛). What do you think? Suggestions and feedback are very welcome!

**Comment by dimateos** [comment]
> 1. Infinite Ground Plane

Sounds nice! Could be used for data overview without need for axes etc.
The idea prompted me to create https://github.com/brenocq/implot3d/discussions/57 and https://github.com/brenocq/implot3d/discussions/58, but I think bounded plots would also benefit from those.

> 2. Free Camera Navigation

I think this would be much less trivial, so I would divide this proposal at least into two separate issues.

*  I agree that for some data, FPS camera navigation might be the best exploration tool.
    *  But maybe in those cases people would already consider exporting the data to other specialized program / 3D tool.
    * Or maybe people might already have 3D scene + camera, potentially integrated with https://github.com/CedricGuillemet/ImGuizmo which apart from transform widgets it has a camera control widget + grid.
* So I see a bit of overlap for some of the potential users of implot3D. Nonetheless for data science users with no camera/scene this could be very useful.

IMO I would delay this 3D FPS camera as I foresee a lot of work and you might be better off focusing on what makes the lib shine first?

**Comment by brenocq** [comment]
Hi @dimateos, I'm converting all feature discussions to issues since GitHub has better support for task tracking (and copilot agent execution) with issues. This discussion was converted into https://github.com/brenocq/implot3d/issues/106
